presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushdownLimitThroughRowNumber.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.LimitNode;
+import io.prestosql.sql.planner.plan.RowNumberNode;
+
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.sql.planner.plan.ChildReplacer.replaceChildren;
+import static io.prestosql.sql.planner.plan.Patterns.limit;
+import static io.prestosql.sql.planner.plan.Patterns.rowNumber;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static java.lang.Math.toIntExact;
+
+public class PushdownLimitThroughRowNumber
+        implements Rule<LimitNode>
+{
+    private static final Pattern<LimitNode> PATTERN = limit()
+            .matching(WindowPushdownRules::pushdownableLimit)
+            .with(source().matching(rowNumber()));
+
+    @Override
+    public Pattern<LimitNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(LimitNode node, Captures captures, Context context)
+    {
+        Lookup lookup = context.getLookup();
+        RowNumberNode source = (RowNumberNode) getOnlyElement(lookup.resolveGroup(node.getSource()).collect(Collectors.toList()));
+        int limit = toIntExact(node.getCount());
+        RowNumberNode rowNumberNode = mergeLimit(source, limit);
+        if (rowNumberNode.getMaxRowCountPerPartition().isPresent() && source.getMaxRowCountPerPartition().isPresent()) {
+            if (rowNumberNode.getPartitionBy().isEmpty()) {
+                return Result.ofPlanNode(rowNumberNode);
+            }
+            // We can stop iterative optimization if the partition count is identical.
```

Comment:
This rule isn't aware of whether the optimizer is "iterative" or some other form, so this comment is not necessarily accurate. Also, it doesn't provide any information that's not readily apparent from the condition below, so let's remove it.